java/core/src/main/java/com/microsoft/bond/BondType.java

```diff
@@ -0,0 +1,514 @@
+     *
+     * @param clazz     the struct class
+     * @param <TStruct> the Bond struct value class
+     * @return type resolver or null
+     */
+    public static <TStruct extends BondSerializable> StructBondType<TStruct> getStructType(
+            Class<TStruct> clazz, BondType<?>... genericTypeArguments) {
+        ArgumentHelper.ensureNotNull(clazz, "clazz");
+        ArgumentHelper.ensureNotNull(genericTypeArguments, "genericTypeArguments");
+        @SuppressWarnings("unchecked")
+        StructBondTypeResolver<TStruct> structTypeResolver =
+                (StructBondTypeResolver<TStruct>) structTypeResolverRegistry.get(clazz);
+        if (structTypeResolver != null) {
+            return structTypeResolver.resolveAndInitialize(genericTypeArguments);
+        } else {
+            return null;
+        }
+    }
+
+    /**
+     * Registers a struct type resolver from generated code.
+     *
+     * @param clazz              the struct class
+     * @param structTypeResolver type resolver for the given struct type
+     * @param <TStruct>          the Bond struct value class
+     */
+    protected static <TStruct extends BondSerializable> void registerStructType(
+            Class<TStruct> clazz, StructBondTypeResolver<TStruct> structTypeResolver) {
+        structTypeResolverRegistry.put(clazz, structTypeResolver);
+    }
+
+    /**
+     * Retrieves type descriptor for the given enum class, or null if none exists.
+     *
+     * @param clazz   the enum class
+     * @param <TEnum> the Bond enum value class
+     * @return type descriptor or null
+     */
+    public static <TEnum extends BondEnum> EnumBondType<TEnum> getEnumType(Class<TEnum> clazz) {
+        ArgumentHelper.ensureNotNull(clazz, "clazz");
+        @SuppressWarnings("unchecked")
+        EnumBondType<TEnum> enumType = (EnumBondType<TEnum>) enumTypeRegistry.get(clazz);
+        return enumType;
+    }
+
+    /**
+     * Registers an enum type from generated code, also caching it in the type cache.
+     *
+     * @param clazz    the enum class
+     * @param enumType type descriptor for the given enum type
+     * @param <TEnum>  the Bond enum value class
+     * @return cached instance of the enum type descriptor
+     */
+    protected static <TEnum extends BondEnum> EnumBondType<TEnum> registerEnumTypeWithCaching(
+            Class<TEnum> clazz, EnumBondType<TEnum> enumType) {
+        EnumBondType<TEnum> cachedEnumType = (EnumBondType<TEnum>) typeCache.get(enumType);
+        enumTypeRegistry.put(clazz, cachedEnumType);
+        return cachedEnumType;
+    }
+
+    /**
+     * Returns a cached type descriptor that is equal to the argument, or caches the argument otherwise.
+     *
+     * @param type the type descriptor
+     * @param <T>  the Bond value class
+     * @return a cached type descriptor
+     */
+    protected static <T> BondType<T> getCachedType(BondType<T> type) {
```

Comment:
I have redesigned this a bit, and simplified type cache.

---
In reply to: [123102269](https://github.com/Microsoft/bond/pull/502#discussion_r123102269) [](ancestors = 123102269)